src/redash_stmo/handlers/query_results.py

```diff
@@ -0,0 +1,52 @@
+from redash_stmo import extract_table_names
+
+from redash.handlers.query_results import QueryResultResource
+from redash.permissions import require_access, require_permission, view_only
+from redash.handlers.base import get_object_or_404
+from redash import models
+
+
+class QueryResultResourceEx(QueryResultResource):
+    @require_permission('view_query')
+    def get(self, query_id=None, query_result_id=None, filetype='json'):
+        if query_result_id:
+            query_result = get_object_or_404(models.QueryResult.get_by_id_and_org, query_result_id, self.current_org)
+            # Look for queries matching this result whose data source is 'Query Results'.
+            if models.Query.query.join(
+                    models.DataSource,
+            ).filter(
+                models.Query.query_hash == query_result.query_hash,
+                models.DataSource.type == 'results',
+            ).first():
+
+                for table_name in extract_table_names.extract_tables(query_result.query_text):
+                    # Look for query IDs being accessed.
+                    if table_name.startswith("query_"):
+                        try:
+                            qid = int(table_name.split('_', 1)[1])
+                        except ValueError:
+                            # If it's not "query_NNN" it can't affect our permissions check here.
+                            continue
+                        upstream_q = models.Query.query.filter(models.Query.id == qid).first()
+                        if upstream_q is None:
+                            continue
+                        # If the user making this request doesn't have permission to
+                        # view the query results being accessed in this query, deny
+                        # access.
+                        require_access(upstream_q.data_source.groups, self.current_user, view_only)
+
+            require_access(query_result.data_source.groups, self.current_user, view_only)
+
+        return QueryResultResource.get(self, query_id, query_result_id, filetype)
+
+def query_results_permissions(app):
+    from redash.handlers import api
+    # remove the original resource
+    del app.view_functions['query_result']
+    api.app = app
```

Comment:
@emtwo added [`redash_stmo.resources.add_resource`](https://github.com/mozilla/redash-stmo/blob/master/src/redash_stmo/resources.py) since the `api.app` setting is repetitive.

Reply:
That calls `add_org_resource` with a different signature.

Reply:
Right, let's update the function to cover both cases by simply pass through the positional and key value arguments:

```python
def add_resource(app, *args, **kwargs):
    """
    After api.init_app() is called, api.app should be set by Flask (but it's not) so that
    further calls to add_resource() are handled immediately for the given app.
    """
    api.app = app
    api.add_org_resource(*args, **kwargs)
```